### ISSUE TITLE
SPA compatibility is gone since v39.19.0

### ISSUE THREAD

**Issue by qtc-de** [jdalrymple/gitbeaker#3427]
**Description**

- Node.js version: v18.18
- Gitbeaker version: v39.19.0
- Gitbeaker release (cli, node, browser, core, requester-utils): requester-utils
- OS & version: alpine

First of all: I have no deep understanding of the Java Script Package system, so please correct me if I got something wrong.

I use gitbeaker within a SPA build with vite & vuejs and it worked really great so far (kudos for this project :heart:). Installation is done via `package.json` which already resulted in the following warning several times:

**package.json**
```json
...
  "dependencies": {
    "@gitbeaker/rest": "^39.12.0",
    ...
}
```

**Build Warning**
```
[plugin:vite:resolve] Module "https" has been externalized for browser compatibility, imported by "/builds/example/node_modules/@gitbeaker/rest/dist/index.mjs". See http://vitejs.dev/guide/troubleshooting.html#module-externalized-for-browser-compatibility for more details.
```

I didn't care because the application worked anyway. However, with commit 2dcd3d8a68ced5b6fee9f506e423115ca0c1d705 the dependencies `async-sema` and `micromatch` were added and these are causing problems when using the package from the browser. First of all, there are way more warnings during build:

```
[plugin:vite:resolve] Module "https" has been externalized for browser compatibility, imported by "/builds/example/node_modules/@gitbeaker/rest/dist/index.mjs". See http://vitejs.dev/guide/troubleshooting.html#module-externalized-for-browser-compatibility for more details.
[plugin:vite:resolve] Module "events" has been externalized for browser compatibility, imported by "/builds/example/node_modules/async-sema/lib/index.js". See http://vitejs.dev/guide/troubleshooting.html#module-externalized-for-browser-compatibility for more details.
[plugin:vite:resolve] Module "util" has been externalized for browser compatibility, imported by "/builds/example/node_modules/micromatch/index.js". See http://vitejs.dev/guide/troubleshooting.html#module-externalized-for-browser-compatibility for more details.
[plugin:vite:resolve] Module "path" has been externalized for browser compatibility, imported by "/builds/example/node_modules/picomatch/lib/utils.js". See http://vitejs.dev/guide/troubleshooting.html#module-externalized-for-browser-compatibility for more details.
[plugin:vite:resolve] Module "path" has been externalized for browser compatibility, imported by "/builds/example/node_modules/picomatch/lib/picomatch.js". See http://vitejs.dev/guide/troubleshooting.html#module-externalized-for-browser-compatibility for more details.
[plugin:vite:resolve] Module "path" has been externalized for browser compatibility, imported by "/builds/example/node_modules/picomatch/lib/constants.js". See http://vitejs.dev/guide/troubleshooting.html#module-externalized-for-browser-compatibility for more details.
[plugin:vite:resolve] Module "util" has been externalized for browser compatibility, imported by "/builds/example/node_modules/fill-range/index.js". See http://vitejs.dev/guide/troubleshooting.html#module-externalized-for-browser-compatibility for more details.
```

The build succeeds, but when opening the app within the browser one gets the error message:

```
Uncaught TypeError: events_1.default is not a constructor
```

I already tried some workarounds, but even after fixing this error message, others appear.

**Steps to reproduce**

Not sure how to reproduce it exactly. I have an already existing (private) project that worked for previous releases but brakes with v39.19.0. I guess simply creating a vite SPA application and installing gitbeaker via `packages.json` could be sufficient to reproduce.

**Expected behaviour**

Installation via `packages.json` should be possible for SPA usage too?!

**Actual behaviour**

Installation is not possible for SPA because some dependency modules are not compatible.

**Possible fixes**

As I said, the problem could be on my side. E.g. your documentation recommends to install gitbeaker via CDN for browser usage. Since my App is an offline SPA, this is not that easy and shipping a local copy is not that elegant than just installing via `package.json`.

I guess the problems are the newly added dependencies and need to be resolved by these projects.

**Checklist**
- [x] I have checked that this is not a duplicate issue.
- [x] I have read the documentation.

**Comment by jdalrymple** [comment]
Hey!

Yea, so the issue looks like the resolution of those other two packages. The reason why the browser version/docs suggest using the CDN is because that CDN actually bundles the library with the dependencies nicely to be resolved in the browser. When just using the node install, those other libraries are not imported correctly.

You could try installing locally from the cdn url?

I have tested the package in the browser (see packages/rest/tests/e2e) without an issue so I know it can be done, i just don't know the particular settings needed for a SPA

**Comment by jdalrymple** [comment]
Ill look into this deeper though, maybe a way for me to make it more browser friendly. For now, perhaps use a previous version as i do some investigations. Sorry for the inconvenience !

**Comment by qtc-de** [comment]
The problem with the CDN solution is that the app is supposed to run in an offline environment. Therefore, I would need to download and cache the script during the build process. Moreover, using CDN components in vue SPA looks quite hacky. Not sure whether where is an elegant way to do it, but all I see so far looks quite complicated. Just adding the dependency to `packages.json` is way cleaner and easier to maintain.

As far as I understand it there is probably not much you can do. I guess the problems are caused by the imported library. I just noticed that there is even an open issue that looks quite similar to my problem: https://github.com/vercel/async-sema/issues/48 and also a proposed solution via a PR: https://github.com/vercel/async-sema/pull/51

Unfortunately, the project seems to be no longer maintained and it is questionable whether the PR ever gets merged :thinking: But maybe you also find another solution. Thanks for your efforts :pray:

**Comment by jdalrymple** [comment]
I mean you can add the dep to the package.json using the cdn url. 

Yea? Ill look at the PR, worst case i fork the repo and merge it

**Comment by qtc-de** [comment]
> I mean you can add the dep to the package.json using the cdn url.

This is possible? Like I said, I have no idea of the whole JavaScript ecosystem. Can you tell me how the syntax looks like?

**Comment by jdalrymple** [comment]
Sorry, that is only git urls they support. Looks like the inclusion is via url imports ie:

```
import confetti from 'https://esm.sh/canvas-confetti@1.6.0'
```

But i believe that happens at runtime which doesnt help you much. For now i suggest using an older version to not be a roadblock for you. Ill see what i can do fix it up for you ASAP

**Comment by trevor-vaughan** [comment]
Ouch. Looks like someone might need to reach out to Vercel directly. That package is used by a LOT of projects.

**Comment by jdalrymple** [comment]
Someone did make a [PR](https://github.com/vercel/async-sema/pull/51) and tagged one of the owners, but no activity thus far from what i can see.

**Comment by trevor-vaughan** [comment]
Hmm....maybe in their GitLab discussions channel? https://github.com/orgs/vercel/discussions

**Comment by jdalrymple** [comment]
I updated the micromatch ref to another lib that is browser friendly, so that should remove a most of those warnings, but the events one really depends on that PR

**Comment by jdalrymple** [comment]
So my latest changes fix the issue you brought up, but this message:

>[plugin:vite:resolve] Module "https" has been externalized for browser compatibility, imported by "/builds/example/node_modules/@gitbeaker/rest/dist/index.mjs". See http://vitejs.dev/guide/troubleshooting.html#module-externalized-for-browser-compatibility for more details.

Still remains. 

Im using a dynamic import that looks for browser usage before doing the import, but i guess when vite is building, its just seeing the presence of the [import](https://github.com/jdalrymple/gitbeaker/blob/main/packages/rest/src/Requester.ts#L25C1-L31C6) and sending a warning. 

```js
    if (typeof window !== 'object') {
      const { Agent } = await import('https');

      options.agent = new Agent({
        rejectUnauthorized: false,
      });
    }
```

**Comment by jdalrymple** [comment]
:rocket: Issue was released in [`39.30.0`](https://github.com/jdalrymple/gitbeaker/releases/tag/39.30.0) :rocket:
